### DISCUSSION TITLE
Update on some work in progress.

### DISCUSSION THREAD

**Discussion by bitbound** [immense/Remotely#257] [Show and tell]
Hey, all.

I thought I'd give a quick update on some stuff I'm working on since it's going to take a while to complete.  I've been playing with Blazor (a C# web framework from Microsoft) off-and-on for a bit, and my latest experiments have inspired me to rewrite Remotely with it.

I'm taking the opportunity to redesign the UI, which I'm hoping will be more appealing and friendly to use.  This will also allow me to implement new features easier.  The current front-end is a jumbled mess that I've carried over from previous iterations of the project, and adding anything to it has become increasingly burdensome.

A proper SPA framework will alleviate a lot of headaches.  I plan on adding a few things right away, such as saved scripts.

Here's a preview of the new UI.

![Remotely Preview](https://i.imgur.com/Y4qNKa5.gif)

**Comment by stevesobol** [comment]
That is "HOLY CRAP THAT'S AMAZING"-level wonderful.

**Comment by kenbshinn** [comment]
That is a sweet interface. I can't wait until it comes out.

**Comment by bmcgonag** [comment]
Excellent looking as always.

**Comment by bitbound** [comment]
Forgot to mention that I added a new, simplified GitHub Actions workflow named "Build".  It doesn't try to deploy.  You just enter the server URL and target server runtime, and it builds a ZIP package.

![image](https://user-images.githubusercontent.com/20995508/112738614-a6333a80-8f21-11eb-9dfa-da806165c48b.png)

**Comment by swilkey** [comment]
Jared
The interface looks really impressive!
Thanks
Stephen

**Comment by PhillipBrickner** [comment]
This looks Awesome!

**Comment by lincoln131** [comment]
I love this!

**Comment by opensource008** [comment]
wow nice work and awesome new ui..
Keep safe and God bless

**Comment by bitbound** [comment]
Just wanted to give a quick update here.

I've got most of the existing functionality re-implemented.  I've done a lot of refactoring and tweaked a few things and hopefully made them easier to use.  There are some new features too, like a tab for remote logs and script history.

There will be a full, detailed list of changes before the actual release, as there are some breaking changes (not too bad, though, and only in some cases).

I'm working on scripts right now, and once I'm done, I'll have a preview release ready to try out.

For scripts, I'm going to add saving them within the app itself.  I plan to add the following in the initial release:
- A page for adding, managing, etc. all your saved scripts.
- A toggle for private vs shared.
- A toggle for adding it to "Quick Scripts" on the terminal.
- A few basic environment variables that will be available within the script (device ID, server URL, etc.). 
- A toggle for automatically creating an alert if there are errors.
- Maybe an option for creating an alert if output contains or doesn't contain a keyword.  Would that be useful?
- Automated script schedules.  I plan to do a basic scheduling system first.  Maybe just start with monthly, weekly, and daily.  What's most important for you guys?  Without going too crazy, like "every Wednesday of every 3rd week of every 2nd month when it's sunny."

Let me know your thoughts.  :)

**Comment by swilkey** [comment]
Hi Jared

This sounds great.  The one area that I think remotely needs to improve is
the upgrade process.  My understanding at present is that if you miss a
version in upgrading you have to do a complete reinstall of the server and
you may need to completely reinstall all client software if there has been
a breaking charge.  Maybe this has been improved already and I haven't
realised it.

Obviously this isn't a big problem if you are just using the instant
support client, but if you are using the agent client then you have to
reregister all the clients to the server after reinstalling the server. If
as we currently do, you keep the client disabled unless it is needed, that
reregistration doesn't happen automatically.

Anyway, I love all the other changes you are making, I just wanted to add
the upgrade process into people's thinking about it.

Stephen

**Reply by bitbound** [reply]
Thanks for the feedback, @swilkey .

The majority of what you're talking about has been resolved.  There's a separate upgrade script now that only puts the new files in place.  However, whether using that or re-running the initial installation script, the agents should be retaining all their registration info after updating themselves.

That said, I don't ever use these scripts since I use an automated build pipeline to deploy new versions.  I'm also using Caddy Server instead of Nginx.  So I mostly rely on people letting me know if the process isn't working correctly.

If you're still experiencing any issues upgrading in one of the recent releases, let me know, and I'll take a look at it.  :)

I plan to revisit the building and deployment processes after the Mac client, though.  I want to get to a place where everyone's server was built specifically for their server, with the URLs embedded in the clients.  Not being able to rely on that has made it challenging to create a good user experience, and it's a limiting factor in some features.

For example, I don't think very many people know that there's an light-weight, installable, self-updating version of the quick support client.  It has to be built from source with the server's URL embedded, or the installer won't even work.  So I hide it on the Downloads page if that's the case.

**Comment by bitbound** [comment]
Another longer preview video.  :)  Scripts is coming along nicely, I think!

https://www.reddit.com/r/remotely_app/comments/mk7x98/another_preview_of_new_ui_and_workinprogress_on/

Edit: Some new things that might be worth pointing out.
- You can now sort by various device properties.
- Remote logs are grabbed on demand from the remote device.
- Script history is new.
- There is no longer the hovering window for auto-complete.  If people were super attached to it, I can try adding it back in. 
- Instead, auto-complete for PowerShell Core executes on the remote device.  This allows you to auto-complete folder paths and such that are specific to the device.  (This works on all client operating systems.)  Ctrl + Space will show all completions instead of cycling through them, with a description in their hovertip.
- Editing device groups is easier.  Just click the checkbox.

**Comment by sinichi19** [comment]
@lucent-sea another awesome work sir.
Kindly advise if you need beta tester
Will try to update on existing remotely and fresh installed.

Thank you and Keep Safe.

**Comment by stevesobol** [comment]
Wow.

I want to move from Connectwise Control to Remotely because Connectwise's branding capabilities don't just suck, they're nonexistent, and the simplest things like providing a custom web page from where a client* can connect to me for a remote session are... impossible.

I put off creating a branded installer because you're moving so quickly, adding lots of features I want to use, but I'm still going to do it.

But this new UI is a bonus. It already looks so much more awesome than the current UI. It's the whipped cream on top of a huge piece of strawberry cheesecake.

Good job!

*client, in this case, means "person paying me for services", not "the client side of a client/server connection" :)

**Comment by sencagri** [comment]
Very nice, looking beautiful. I have some experience with Blazor. I will download and check if there are anything I can do.

**Comment by bitbound** [comment]
I'm not fully done testing and refining, but if anyone wants to try it out, it should be stable enough.  The server code itself, anyway.  The server installer tool is brand new, and I haven't tested it much yet.

The changes are also live on https://app.remotely.one if you want to try it there.

Be sure to back up everything before attempting to upgrade.  The database migrations were seamless on my test server and public server, but you never know.

Let me know how it goes!  :)

**BREAKING CHANGES ALERT!  Please read all of the below.  :)**

- There's a new Server Installer CLI tool that will show on the releases page instead of the install scripts and ZIP packages.
  - It integrates with GitHub's REST API to build a customized server using GitHub Actions, with your server's URL embedded in the clients.
  - Instructions for using it are now in the readme.
- The default reverse proxy going forward is Caddy Server.  But you can still select Nginx when using the above-mentioned installer tool.
- The `Downloads` folder has been renamed to `Content`.  I've created a redirect in the server code for file downloads, though, so it should be a seamless transition.  But just be aware that they live there now.
    -  I won't go into all the OCD details about why I renamed it.  But you can be rest assured that it was 100% not necessary, and I'm sorry.
 - API keys now use the key's `Id` instead of a separate `Token` property/column.  It didn't make sense to have two separate GUID columns, and it was confusing.  If you're using API tokens, you'll need to update your integrations so they use the GUID in the `Id` column, as the `Token` column will now be gone.
- The Nginx websocket route for `BrowserHub` has changed to `_blazor`.
- Some of the API controller routes changed.  Some new ones were added.  Check the Swagger page.
- File uploads and downloads now use an expiring authentication token instead of the files themselves expiring.  This happens behind-the-scenes, so you only need to worry about it if you had an integration that used the FileSharing.  You can also authorize with a valid API token.

![image](https://user-images.githubusercontent.com/20995508/114649147-cf2f2b80-9c94-11eb-8625-d45e45041f76.png)

**Comment by bitbound** [comment]
Also, I plan on doing a full start-to-finish video on how to set up a server.

**Reply by ajshea** [reply]
It works! 
Had to pay close attention to the video, and review several points. 
TLS cert didn't install on nginx, but running letsencrypt took care of that very nicely.

**Comment by bitbound** [comment]
Oh.  I also forgot to mention.  I've made some adjustments to how image quality and output throttling are automatically adjusted based on available bandwidth.  If you notice any difference, good or bad, let me know.  :)

**Comment by opensource008** [comment]
@lucent-sea  

Good Day,

Want only to share Sir the test I made, I already updated my test server using your new version (v2021.04.29.0623 Release)
Working without problem

4core CPU and 6gb memory
Sqlite database
Ubuntu 20.04
Nginx w/ SSL
Installed 100 agents (physical computers)
Tried using Portable Clients.

Noticed the system is more light (server side)

Next need to try is the response of the server if there is 100 agents online (physical computers) and restart the server, 
because from the previous version when there is many computers are online and restart the server, experienced 100%memory usage then hang-up dont know if there is problem from the remotely system to database coming from boot up and need to read many online agents,.

Thank you Sir for your wonderful work, More power and God Bless.
